tests/security/SingleParticipantWithSecurity/pubsub.cpp

```diff
@@ -0,0 +1,266 @@
+                                  TOPIC_QOS_DEFAULT,
+                                  DDS::TopicListener::_nil(),
+                                  OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (!topic) {
+        ACE_ERROR_RETURN((LM_ERROR,
+                          ACE_TEXT("%N:%l: main()")
+                          ACE_TEXT(" ERROR: create_topic failed!\n")),
+                         EXIT_FAILURE);
+      }
+
+      // Create Publisher
+      DDS::Publisher_var pub =
+        participant->create_publisher(PUBLISHER_QOS_DEFAULT,
+                                      DDS::PublisherListener::_nil(),
+                                      OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (!pub) {
+        ACE_ERROR_RETURN((LM_ERROR,
+                          ACE_TEXT("%N:%l: main()")
+                          ACE_TEXT(" ERROR: create_publisher failed!\n")),
+                         EXIT_FAILURE);
+      }
+
+      DDS::DataWriterQos qos;
+      pub->get_default_datawriter_qos(qos);
+      if (dw_reliable()) {
+        std::cout << "Reliable DataWriter" << std::endl;
+        qos.history.kind = DDS::KEEP_ALL_HISTORY_QOS;
+        qos.reliability.kind = DDS::RELIABLE_RELIABILITY_QOS;
+      }
+
+      // Create DataWriter
+      DDS::DataWriter_var dw =
+        pub->create_datawriter(topic.in(),
+                               qos,
+                               DDS::DataWriterListener::_nil(),
+                               OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (!dw) {
+        ACE_ERROR_RETURN((LM_ERROR,
+                          ACE_TEXT("%N:%l: main()")
+                          ACE_TEXT(" ERROR: create_datawriter failed!\n")),
+                         EXIT_FAILURE);
+      }
+
+      // Create Subscriber
+      DDS::Subscriber_var sub =
+        participant->create_subscriber(SUBSCRIBER_QOS_DEFAULT,
+          DDS::SubscriberListener::_nil(),
+          OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (!sub) {
+        ACE_ERROR_RETURN((LM_ERROR,
+          ACE_TEXT("%N:%l main()")
+          ACE_TEXT(" ERROR: create_subscriber() failed!\n")),
+          EXIT_FAILURE);
+      }
+
+      // Create DataReader
+      DataReaderListenerImpl* const listener_servant = new DataReaderListenerImpl;
+      DDS::DataReaderListener_var listener(listener_servant);
+
+      DDS::DataReaderQos dr_qos;
+      sub->get_default_datareader_qos(dr_qos);
+      if (DataReaderListenerImpl::is_reliable()) {
+        std::cout << "Reliable DataReader" << std::endl;
+        dr_qos.reliability.kind = DDS::RELIABLE_RELIABILITY_QOS;
+      }
+
+      DDS::DataReader_var reader =
+        sub->create_datareader(topic.in(),
+          dr_qos,
+          listener.in(),
+          OpenDDS::DCPS::DEFAULT_STATUS_MASK);
+
+      if (!reader) {
+        ACE_ERROR_RETURN((LM_ERROR,
+          ACE_TEXT("%N:%l main()")
+          ACE_TEXT(" ERROR: create_datareader() failed!\n")),
+          EXIT_FAILURE);
+      }
+
+      // Start writing threads
```

Comment:
From here, the flow should be
1. Wait for the writer to match the reader.
2. Write the sample.
3. Read the sample.

You shouldn't need to spawn a thread for the writer or use a listener for the reader.  You may need a waitset for the reader or you can use one of the test libraries (`wait_for_sample`, maybe).